lib/Sema/CSApply.cpp

```diff
@@ -6930,8 +6915,28 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
 
       maybeDiagnoseUnsupportedFunctionConversion(cs, expr, toFunc);
 
-      return cs.cacheType(new (tc.Context)
-                              FunctionConversionExpr(expr, toType));
+      // SWIFT_ENABLE_TENSORFLOW
+      auto toEINoAdConversion =
+          toEI.withDifferentiable(fromEI.isDifferentiable());
+      auto toFuncNoADConversion = toFunc->withExtInfo(toEINoAdConversion);
+      expr = cs.cacheType(new (tc.Context)
+                              FunctionConversionExpr(expr,
+                                                     toFuncNoADConversion));
+
+      // Make the conversion to @autodiff happen after all other conversions,
+      // because some of the other conversions are not currently supported on
+      // @autodiff functions. (e.g. escape_to_noescape).
+      // After we do support those conversions, the order will no longer matter.
+      if (!fromEI.isDifferentiable() && toEI.isDifferentiable()) {
+        expr = cs.cacheType(new (tc.Context)
+                            AutoDiffFunctionExpr(expr, toFunc));
+      }
+      else if (fromEI.isDifferentiable() && !toEI.isDifferentiable()) {
```

Comment:
If we have an `@escaping @autodiff` function and we need to convert it to a *non-escaping non-differentiable* function, won't this emit a `(autodiff_function_extract_original_expr (convert_function_expr f))` which would then trigger a reference counting crasher again? In this case, I think `autodiff_function_extract_original` should happen before `function_conversion_expr`.